tracing-appender/src/rolling.rs

```diff
@@ -558,17 +569,73 @@ impl Inner {
                     .unwrap_or(0),
             ),
             rotation,
+            keep_last,
         };
         Ok((inner, writer))
     }
 
+    fn prune_old_logs(&self, keep_last: usize) {
+        let files = fs::read_dir(&self.log_directory).map(|dir| {
+            dir.filter_map(Result::ok)
+                .filter(|entry| {
+                    entry.file_name().to_string_lossy().contains(
+                        &self
+                            .log_filename_prefix
+                            .clone()
+                            .expect("prefix is always present"),
+                    )
+                })
+                .collect::<Vec<_>>()
+        });
+        let mut files = match files {
+            Ok(files) => files,
+            Err(error) => {
+                eprintln!("Error reading the log directory/files: {}", error);
+                return;
+            }
+        };
+        if files.len() < keep_last {
+            return;
+        }
+        // sort the files by their creation timestamps.
+        files.sort_by_key(|file| {
+            file.metadata()
+                .and_then(|metadata| metadata.created().map(Some))
+                .unwrap_or_else(|error| {
+                    eprintln!(
+                        "Unable to get file creation time for {}: {}",
+                        file.path().display(),
+                        error
+                    );
+                    // if we couldn't determine the file's creation timestamp
+                    // for whatever reason, it should compare as "less than"
+                    // other files, so that it's not deleted.
+                    None
+                })
```

Comment:
i think your comment https://github.com/tokio-rs/tracing/pull/2284#discussion_r977347938 was correct and we probably shouldn't just assume that any log files we can't get the creation time of are "infinitely old",  because then we might end up deleting log files that are actually newer. instead, if we encounter an error getting any file's creation timestamp, this method should probably just print the error and return immediately so that we don't delete anything.